shared/js/content-scope/utils.js

```diff
@@ -0,0 +1,156 @@
+        }
+    }
+}
+
+export function isFeatureBroken (args, feature) {
+    return args.site.brokenFeatures.includes(feature)
+}
+
+/**
+ * For each property defined on the object, update it with the target value.
+ */
+export function overrideProperty (name, prop) {
+    // Don't update if existing value is undefined or null
+    if (!(prop.origValue === undefined)) {
+        /**
+         * When re-defining properties, we bind the overwritten functions to null. This prevents
+         * sites from using toString to see if the function has been overwritten
+         * without this bind call, a site could run something like
+         * `Object.getOwnPropertyDescriptor(Screen.prototype, "availTop").get.toString()` and see
+         * the contents of the function. Appending .bind(null) to the function definition will
+         * have the same toString call return the default [native code]
+         */
+        try {
+            defineProperty(prop.object, name, {
+                get: () => JSON.stringify(prop.targetValue)
+            })
+        } catch (e) {
+        }
+    }
+    return prop.origValue;
+}
+
+// TODO make rollup aware of this so it can tree shake
+const mozProxies = 'wrappedJSObject' in window
+
+export function defineProperty (object, propertyName, descriptor) {
+    if (mozProxies) {
+        const usedObj = object.wrappedJSObject
+        const UsedObjectInterface = window.wrappedJSObject.Object
+        const definedDescriptor = new UsedObjectInterface();
+        ['configurable', 'enumerable', 'value', 'writable'].forEach((propertyName) => {
+            // TODO check if value is complex and export it if so.
+            if (propertyName in descriptor) {
+                definedDescriptor[propertyName] = descriptor[propertyName]
+            }
+        });
+        ['get', 'set'].forEach((methodName) => {
+            if (methodName in descriptor) {
+                exportFunction(descriptor[methodName], definedDescriptor, { defineAs: methodName })
+            }
+        })
+        UsedObjectInterface.defineProperty(usedObj, propertyName, definedDescriptor)
+    } else {
+        Object.defineProperty(object, propertyName, descriptor)
+    }
+}
+
+export class DDGProxy {
+    constructor (objectScope, property, proxyObject) {
+        if (mozProxies) {
+            this._native = objectScope[property]
+            const handler = new window.wrappedJSObject.Object()
+            handler.apply = exportFunction(proxyObject.apply, window)
+            this.internal = new window.wrappedJSObject.Proxy(objectScope.wrappedJSObject[property], handler)
+            exportFunction(this.internal, objectScope, { defineAs: property })
+        } else {
+            this._native = objectScope[property]
+            const handler = {}
+            handler.apply = proxyObject.apply
+            this.internal = new window.Proxy(objectScope[property], handler)
+            objectScope[property] = this.internal
+        }
+    }
+}
```

Comment:
Does this need to be a class? We never assign this to a variable, nor is there any class method which we might want to use? I also don't see the object properties being used.

Reply:
this._native is used in canvas

Reply:
Added the mentioned second step `override()` method to make the "new" not have side effects and that we always use the proxy in a variable.